File: 算法/代码随想录算法PDF/00调试源代码/383.赎金信.cpp
/*
 * @lc app=leetcode.cn id=383 lang=cpp
 *
 * [383] 赎金信
 */

// @lc code=start
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        vector<int> charMap(26, 0);  // 26个字母的出现次数
        if (ransomNote.size() > magazine.size()) {  // 可有可无，有的话快一些
            return false;
        }
        for (int i = 0; i < magazine.size(); i++) {
            charMap[magazine[i] - 'a']++;
        }
        for (int i = 0; i < ransomNote.size(); i++) {
            charMap[ransomNote[i] - 'a']--;
            if (charMap[ransomNote[i] - 'a'] < 0) {
                return false;
            }
        }
        return true;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/383.赎金信.py
#
# @lc app=leetcode.cn id=383 lang=python3
#
# [383] 赎金信
#

# @lc code=start
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        charMap = [0] * 26  # 字符哈希数组
        for ch in magazine:
            charMap[ord(ch) - ord('a')] += 1
        for ch in ransomNote:
            charMap[ord(ch) - ord('a')] -= 1
            if charMap[ord(ch) - ord('a')] < 0:
                return False
        return True
# @lc code=end

